src/lang/parser_helper.ml

```diff
@@ -168,16 +176,51 @@ let mk_fun ~pos args body =
   let fv = Term.free_vars ~bound body in
   mk ~pos (Fun (fv, args, body))
 
-let mk_let ~pos (doc, replace, pat, def) body =
-  mk ~pos (Let { doc; replace; pat; gen = []; def; body })
-
 let mk_rec_fun ~pos pat args body =
-  let name = match pat with PVar [name] -> name | _ -> assert false in
+  let name =
+    match pat with
+      | PVar l when l <> [] -> List.hd (List.rev l)
+      | _ -> assert false
+  in
   let bound = List.map (fun (_, x, _, _) -> x) args in
   let bound = name :: bound in
   let fv = Term.free_vars ~bound body in
   mk ~pos (RFun (name, fv, args, body))
 
+let mk_eval ~pos (doc, pat, def, body) =
+  let ty = Type.var () in
```

Comment:
Didn't try it (I thought of it after my counter-example) but I think that
```suggestion
  let ty = Type.var ~level:(-1) () in
```
might work since it will prevent type generalizations...